neard/src/runtime.rs

```diff
@@ -48,6 +49,42 @@ const POISONED_LOCK_ERR: &str = "The lock was poisoned.";
 const STATE_DUMP_FILE: &str = "state_dump";
 const GENESIS_ROOTS_FILE: &str = "genesis_roots";
 
+/// Wrapper type for epoch manager to get avoid implementing trait for foreign types.
+pub struct SafeEpochManager(pub Arc<RwLock<EpochManager>>);
+
+impl AsRef<RwLock<EpochManager>> for SafeEpochManager {
+    fn as_ref(&self) -> &RwLock<EpochManager> {
+        self.0.as_ref()
+    }
+}
+
+impl EpochInfoProvider for SafeEpochManager {
+    fn get_validator_stake(
+        &self,
+        epoch_id: &EpochId,
+        last_block_hash: &CryptoHash,
+        account_id: &AccountId,
+    ) -> Result<Option<Balance>, EpochError> {
+        let mut epoch_manager = self.0.write().expect(POISONED_LOCK_ERR);
+        let slashed = epoch_manager.get_slashed_validators(last_block_hash)?.clone();
+        if slashed.contains_key(account_id) {
+            return Ok(None);
+        }
+        let epoch_info = epoch_manager.get_epoch_info(&epoch_id)?;
+        if let Some(index) = epoch_info.validator_to_index.get(account_id) {
+            Ok(Some(epoch_info.validators[*index as usize].stake))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn validator_total_stake(&self, epoch_id: &EpochId) -> Result<Balance, EpochError> {
+        let mut epoch_manager = self.0.write().expect(POISONED_LOCK_ERR);
+        let epoch_info = epoch_manager.get_epoch_info(&epoch_id)?;
+        Ok(epoch_info.validators.iter().map(|info| info.stake).sum())
```

Comment:
What about slashed?

Reply:
I think it doesn't make sense for total stake to change within an epoch.